velox/connectors/hive/HiveDataSink.h

```diff
@@ -129,16 +191,97 @@ class HiveDataSink : public DataSink {
   void close() override;
 
  private:
-  std::unique_ptr<dwrf::Writer> createWriter();
+  std::unique_ptr<dwrf::Writer> createWriter(
+      const std::shared_ptr<const HiveWriterParameters>& writerParameters)
+      const;
+
+  // Move the existing writers to the new indices, if the current input causes
+  // the partitionIdGenerator_ to rehash that changes the IDs of the previous
+  // partition values and the corresponding writers.
+  void moveWriters();
+
+  static RowVectorPtr makeWriterInput(
+      const RowVectorPtr& rawInput,
+      vector_size_t size,
+      const BufferPtr& index);
+
+  // Compute the count of rows as well as the concrete row indices for every
+  // partition ID based on the ID labeling of rowPartitionIds. Both rowCounts
+  // and rowIndices are indexed by partition ID. rowIndices[i] is a vector of
+  // all rows labeled with partition ID i.
+  static void computePartitionRowCountsAndIndices(
+      const raw_vector<uint64_t>& rowPartitionIds,
+      const SelectivityVector& activeRows,
+      uint64_t maxPartitionId,
+      memory::MemoryPool* FOLLY_NONNULL pool,
+      std::vector<vector_size_t>& rowCounts,
+      std::vector<BufferPtr>& rowIndices);
+
+  RowVectorPtr makeInputPartitionsVector(const RowVectorPtr& input) const;
+
+  std::vector<BaseVector::CopyRange> convertToIncrementalRanges(
+      std::vector<vector_size_t>& rowCounts,
+      std::vector<BufferPtr>& rowIndices) const;
+
+  static std::string getPartitionString(
```

Comment:
This method can be a function in anonymous namespace in the .cpp file.

Reply:
Moved